apps/settings/js/modules/bluetooth/bluetooth_connection_manager.js

```diff
@@ -0,0 +1,674 @@
+        return Promise.reject('disconnect failed');
+      });
+    },
+
+    /**
+     * The method will disconnect the input device with the device's adapter.
+     *
+     * @access public
+     * @memberOf BluetoothConnectionManager
+     * @param {String} address
+     * @returns {Promise} resolve: profiles
+     * @returns {Promise} reject: reason
+     */
+    fetchDeviceConnectionProfiles: 
+    function btcm_fetchDeviceConnectionProfiles(address) {
+      return new Promise((resolve, reject) => {
+        if (!this._defaultAdapter) {
+          reject('default adapter is not existed!!');
+        }
+
+        // Get connected devices from platform side.
+        this.getConnectedDevices().then((connectedDeviceItemsMap) => {
+          var profiles = connectedDeviceItemsMap[address].connectedProfiles;
+          Debug('fetchDeviceConnectionProfiles(): resolve with profiles = ' + 
+                JSON.stringify(profiles));
+          resolve(profiles);
+        }, (reason) => {
+          Debug('fetchDeviceConnectionProfiles(): ' + 
+                'reject with reason: ' + reason);
+          reject(reason);
+        });
+      });
+    },
+
+    /**
+     * The method will get all connected devices profiles 
+     * which we are interested in. Profile: HFP, A2DP.
+     *
+     * @access private
+     * @memberOf BluetoothConnectionManager
+     * @returns {Promise} resolve: connectedDevices
+     * @returns {Promise} reject: reason
+     */
+    getConnectedDevices: function btcm_getConnectedDevices() {
+      // Get connected device via profiles HFP, A2DP
+      return Promise.all([
+        this._getConnectedDevicesByProfile(this.Profiles.HFP),
+        this._getConnectedDevicesByProfile(this.Profiles.A2DP)
+      ]).then((connectedDevices) => {
+        // Update each connected devices in map.
+        var collectedConnectedDevicesByProfile = {
+          'HFP': connectedDevices[0],
+          'A2DP': connectedDevices[1]
+        };
+        var connectedDeviceItemsMap = 
+          this._constructDeviceItemsMap(collectedConnectedDevicesByProfile);
+        Debug('getConnectedDevices(): connectedDeviceItemsMap = ' + 
+              JSON.stringify(connectedDeviceItemsMap));
+        return Promise.resolve(connectedDeviceItemsMap);
+      });
+    },
+
+    /**
+     * The method will get connected device by inputed profile.
+     *
+     * @access private
+     * @memberOf BluetoothConnectionManager
+     * @param {String} profileID
+     * @returns {Promise} resolve: the connected devices in array
+     * @returns {Promise} reject: reason
+     */
+    _getConnectedDevicesByProfile: 
+    function btcm__getConnectedDevicesByProfile(profileID) {
+      return new Promise((resolve, reject) => {
+        if (!this._defaultAdapter) {
+          reject('default adapter is not existed!!');
+        }
+
+        var req = this._defaultAdapter.getConnectedDevices(profileID);
```

Comment:
Because the cost of calling to `getConnectedDevices` is high, gaia should cache the connected devices of each profile. We can achieve this by calling to `getConnectedDevices` in `_initProperties` and update the cache using hfp and a2dp status change events.